Assignment conversion not possible from type "java.lang.Object" to type "net.betzel.fury.test.Object"

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-fury/issues) and found no similar issues.


### Version

0.5.0-SNAPSHOT b573ac651413520b5153d12380d227cef1b911a9

### Component(s)

Java

### Minimal reproduce step

https://github.com/Maurice-Betzel/net.betzel.fury.test

### What did you expect to see?

No compiler exception 

### What did you see instead?

Exception in thread "main" java.lang.RuntimeException: Create compatible serializer failed, 
class: class net.betzel.fury.test.Order
...
Caused by: org.apache.fury.shaded.org.codehaus.commons.compiler.CompileException: File 'net/betzel/fury/test/OrderFuryRefCompatibleCodec_1_414493378_1641313620.java', Line 1602, Column 127: Assignment conversion not possible from type "java.lang.Object" to type "net.betzel.fury.test.Object"


### Anything Else?

CodegenContext needs class-loader fix at line 304:

if (beanClassClassLoader == null) beanClassClassLoader = Fury.class.getClassLoader();

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/codegen/CodegenContext.java
+++ b/java/fury-core/src/main/java/org/apache/fury/codegen/CodegenContext.java
@@ -154,11 +154,8 @@ public class CodegenContext {
 
   public CodegenContext() {}
 
-  public CodegenContext(LinkedHashSet<String> imports) {
-    this.imports = imports;
-  }
-
-  public CodegenContext(Set<String> valNames, LinkedHashSet<String> imports) {
+  public CodegenContext(String pkg, Set<String> valNames, LinkedHashSet<String> imports) {
+    this.pkg = pkg;
     this.valNames = valNames;
     this.imports = imports;
   }
@@ -280,7 +277,7 @@ public class CodegenContext {
    */
   public String type(Class<?> clz) {
     if (!sourcePkgLevelAccessible(clz)) {
-      return "Object";
+      clz = Object.class;
     }
     if (clz.isArray()) {
       return getArrayType(clz);
@@ -292,15 +289,16 @@ public class CodegenContext {
         boolean hasPackage = StringUtils.isNotBlank(pkg);
         Map<String, Boolean> packageMap =
             nameConflicts.computeIfAbsent(hasPackage ? pkg : "", p -> new ConcurrentHashMap<>());
+        Class<?> c = clz;
         Boolean conflictRes =
             packageMap.computeIfAbsent(
                 simpleName,
                 sn -> {
                   try {
                     ClassLoader beanClassClassLoader =
-                        clz.getClassLoader() == null
+                        c.getClassLoader() == null
                             ? Thread.currentThread().getContextClassLoader()
-                            : clz.getClassLoader();
+                            : c.getClassLoader();
                     beanClassClassLoader.loadClass(hasPackage ? pkg + "." + sn : sn);
                     return Boolean.TRUE;
                   } catch (ClassNotFoundException e) {
@@ -337,6 +335,10 @@ public class CodegenContext {
    */
   public void setPackage(String pkg) {
     this.pkg = pkg;
+  }
+
+  public String getPackage() {
+    return pkg;
   }
 
   public Set<String> getValNames() {
@@ -575,7 +577,7 @@ public class CodegenContext {
       List<String> initCodes;
       if (isStatic) {
         if (staticInitCtx == null) {
-          staticInitCtx = new CodegenContext(valNames, imports);
+          staticInitCtx = new CodegenContext(pkg, valNames, imports);
         }
         ctx = staticInitCtx;
         initCodes = staticInitCodes;
@@ -585,7 +587,7 @@ public class CodegenContext {
         }
       } else {
         if (instanceInitCtx == null) {
-          instanceInitCtx = new CodegenContext(valNames, imports);
+          instanceInitCtx = new CodegenContext(pkg, valNames, imports);
         }
         ctx = instanceInitCtx;
         initCodes = instanceInitCodes;

--- a/java/fury-core/src/main/java/org/apache/fury/codegen/ExpressionOptimizer.java
+++ b/java/fury-core/src/main/java/org/apache/fury/codegen/ExpressionOptimizer.java
@@ -113,7 +113,8 @@ public class ExpressionOptimizer {
             });
     // copy variable names so that to avoid new variable name conflict with generated class
     // instance field name.
-    CodegenContext codegenContext = new CodegenContext(ctx.getValNames(), ctx.getImports());
+    CodegenContext codegenContext =
+        new CodegenContext(ctx.getPackage(), ctx.getValNames(), ctx.getImports());
     for (Reference reference : cutExprMap.values()) {
       Preconditions.checkArgument(codegenContext.containName(reference.name()));
     }

--- a/java/fury-core/src/test/java/org/apache/fury/builder/ObjectCodecBuilderTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/builder/ObjectCodecBuilderTest.java
@@ -38,6 +38,7 @@ import org.apache.fury.FuryTestBase;
 import org.apache.fury.codegen.CodeGenerator;
 import org.apache.fury.codegen.CompileUnit;
 import org.apache.fury.codegen.JaninoUtils;
+import org.apache.fury.codegen.javalangnameconflict.MethodSpiltObject;
 import org.apache.fury.config.Language;
 import org.apache.fury.serializer.collection.CollectionSerializersTest;
 import org.apache.fury.test.bean.AccessBeans;
@@ -185,5 +186,15 @@ public class ObjectCodecBuilderTest extends FuryTestBase {
     Fury fury = Fury.builder().requireClassRegistration(false).build();
     AccessBeans.PublicClass object = AccessBeans.createPublicClassObject();
     serDeCheckSerializer(fury, object, "Codec");
+  }
+
+  @Test
+  public void testTypeConflictWhenMethodSplits() {
+    // For issue #1370
+    Fury fury = Fury.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    fury.serialize(new MethodSpiltObject());
+    Assert.assertTrue(
+        fury.getClassResolver().getSerializer(MethodSpiltObject.class)
+            instanceof Generated.GeneratedSerializer);
   }
 }

--- a/java/fury-core/src/test/java/org/apache/fury/codegen/javalangnameconflict/MethodSpiltObject.java
+++ b/java/fury-core/src/test/java/org/apache/fury/codegen/javalangnameconflict/MethodSpiltObject.java
@@ -1,3 +1,22 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
 package org.apache.fury.codegen.javalangnameconflict;
 
 import java.util.List;

--- a/java/fury-core/src/test/java/org/apache/fury/codegen/javalangnameconflict/MethodSpiltObject.java
+++ b/java/fury-core/src/test/java/org/apache/fury/codegen/javalangnameconflict/MethodSpiltObject.java
@@ -1,0 +1,13 @@
+package org.apache.fury.codegen.javalangnameconflict;
+
+import java.util.List;
+import org.apache.fury.test.bean.BeanA;
+
+/** A class used to test `Object` type conflict in #1370. */
+public class MethodSpiltObject extends BeanA {
+  private Object f1;
+  public Object f2;
+  private java.lang.Object f3;
+  private List<java.lang.Object> f4;
+  private List<Object> f5;
+}

